### ISSUE TITLE
Results of GPT2 (124M) reproduction with LitGPT

### ISSUE THREAD

**Issue by awaelchli** [Lightning-AI/litgpt#1607]
The goal of this experiment is to reproduce the loss curves of the original [nanoGPT](https://github.com/karpathy/nanoGPT), but with the [LitGPT](https://github.com/Lightning-AI/litgpt) library. The code is equivalent in most aspects. However, these are notable differences in LitGPT:

- Uses RoPE positional embeddings from Llama's implementation. The original nanoGPT uses different positional embeddings.
- Data in LitGPT is sampled without replacement, nanoGPT used sampling with replacement. LitGPT also uses distributed sampling (via LitData). 

In addition, the LitGPT code was modified to use the Tiktoken tokenizer (not supported by default). Apart from some LitData fixes, no other changes were needed, and the exact hyperparameters were copied from nanoGPT, specifically the [GPT 124M experiment settings](https://github.com/karpathy/nanoGPT/blob/master/config/train_gpt2.py).

## Configuration details

- Branch: [training/gpt2](https://github.com/Lightning-AI/litgpt/compare/main...training/gpt2)
- Lightning 2.4.0dev (bdafe5e7), LitData 0.2.17 (c58b673), PyTorch 2.4
- Hardware: 8xA100 40GB SXM (same as original)

**Commands:**

```commandline
pip install datasets tiktoken
litgpt pretrain --config config_hub/pretrain/nano_gpt.yaml
```

**Full config file:**

```yaml
# config_hub/pretrain/nano_gpt.yaml

# The name of the model to pretrain. Choose from names in ``litgpt.config``. Mutually exclusive with
# ``model_config``. (type: Optional[str], default: null)
model_name: gpt2-124M


# A ``litgpt.Config`` object to define the model architecture. Mutually exclusive with
# ``model_config``. (type: Optional[Config], default: null)
model_config:
  name: gpt2-124M
  block_size: 1024
  vocab_size: 50257
  padded_vocab_size: 50304
  n_layer: 12
  n_head: 12
  n_embd: 768
  head_size: 64
  norm_class_name: LayerNorm
  mlp_class_name: GptNeoxMLP
  bias: false
  lm_head_bias: false
  norm_eps: 1e-5




# Directory in which to save checkpoints and logs. If running in a Lightning Studio Job, look for it in
# /teamspace/jobs/<job-name>/share. (type: <class 'Path'>, default: out/pretrain)
out_dir: out/pretrain/nano-gpt


# The precision to use for pretraining. Possible choices: "bf16-true", "bf16-mixed", "32-true". (type: Optional[str], default: null)
precision: bf16-mixed


# Optional path to a checkpoint directory to initialize the model from.
# Useful for continued pretraining. Mutually exclusive with ``resume``. (type: Optional[Path], default: null)
initial_checkpoint_dir:


# Path to a checkpoint directory to resume from in case training was interrupted, or ``True`` to resume
# from the latest checkpoint in ``out_dir``. An error will be raised if no checkpoint is found. Passing
# ``'auto'`` will resume from the latest checkpoint but not error if no checkpoint exists.
# (type: Union[bool, Literal["auto"], Path], default: False)
resume: false


# Data-related arguments. If not provided, the default is ``litgpt.data.TinyLlama``.
data:
  class_path: litgpt.data.OpenWebText
  init_args:
    data_path: data/openwebtext/optimized
    seed: 2357


# Training-related arguments. See ``litgpt.args.TrainArgs`` for details
train:


  # Number of optimizer steps between saving checkpoints (type: Optional[int], default: 1000)
  save_interval: 4000


  # Number of iterations between logging calls (type: int, default: 1)
  log_interval: 10


  # Number of samples between optimizer steps across data-parallel ranks (type: int, default: 512)
  global_batch_size: 480


  # Number of samples per data-parallel rank (type: int, default: 4)
  micro_batch_size: 12


  # Number of iterations with learning rate warmup active (type: int, default: 2000)
  lr_warmup_steps: 2000


  # Number of epochs to train on (type: Optional[int], default: null)
  epochs:


  # Total number of tokens to train on (type: Optional[int], default: 3000000000000)
  max_tokens: 300_000_000_000


  # Limits the number of optimizer steps to run. (type: Optional[int], default: null)
  max_steps:


  # Limits the length of samples. Off by default (type: Optional[int], default: null)
  max_seq_length: 1024


  # Whether to tie the embedding weights with the language modeling head weights. (type: Optional[bool], default: False)
  tie_embeddings: true


  #   (type: Optional[float], default: 1.0)
  max_norm: 1.0


  #   (type: float, default: 4e-05)
  min_lr: 6e-5


# Evaluation-related arguments. See ``litgpt.args.EvalArgs`` for details
eval:


  # Number of optimizer steps between evaluation calls (type: int, default: 1000)
  interval: 1000


  # Number of tokens to generate (type: Optional[int], default: null)
  max_new_tokens:


  # Number of iterations (type: int, default: 100)
  max_iters: 100


  # Whether to evaluate on the validation set at the beginning of the training
  initial_validation: true


  # Whether to evaluate on the validation set at the end the training
  final_validation: false


# Optimizer-related arguments
optimizer:


  class_path: torch.optim.AdamW
  
  init_args:
    
    #   (type: float, default: 0.001)
    lr: 6e-4
    
    #   (type: float, default: 0.01)
    weight_decay: 0.1
    
    #   (type: tuple, default: (0.9,0.999))
    betas:
      - 0.9
      - 0.95


# How many devices/GPUs to use. Uses all GPUs by default. (type: Union[int, str], default: auto)
devices: auto


# How many nodes to use. (type: int, default: 1)
num_nodes: 1


# Optional path to the tokenizer dir that was used for preprocessing the dataset. Only some data
# module require this. (type: Optional[Path], default: null)
tokenizer_dir: checkpoints/meta-llama/Llama-2-7b-hf


# The name of the logger to send metrics to. (type: Literal['wandb', 'tensorboard', 'csv'], default: tensorboard)
logger_name: wandb


# The random seed to use for reproducibility. (type: int, default: 42)
seed: 42

```

# Training metrics

[**View the full WandB logs here.**](https://api.wandb.ai/links/lightning-oss/t867kuyq)

<img width="1095" alt="image" src="https://github.com/user-attachments/assets/f468663b-9c31-4337-b932-7031ac67a36d">

<img width="1095" alt="image" src="https://github.com/user-attachments/assets/318e784e-9d09-4cd2-8d8c-fd4abe1baf38">

<img width="1088" alt="image" src="https://github.com/user-attachments/assets/aa452cea-6056-415f-ba70-e69e0a8fd9ef">

Last training log lines:

```
Epoch 34 | iter 3050700 step 610140 | loss train: 2.846, val: 2.861 | iter time: 112.20 ms (step) remaining time: 0:01:54
Epoch 34 | iter 3050750 step 610150 | loss train: 2.750, val: 2.861 | iter time: 111.35 ms (step) remaining time: 0:01:49
Epoch 34 | iter 3050800 step 610160 | loss train: 2.815, val: 2.861 | iter time: 110.35 ms (step) remaining time: 0:01:43
Epoch 34 | iter 3050850 step 610170 | loss train: 2.860, val: 2.861 | iter time: 112.53 ms (step) remaining time: 0:01:38
Epoch 34 | iter 3050900 step 610180 | loss train: 2.909, val: 2.861 | iter time: 111.32 ms (step) remaining time: 0:01:32
Epoch 34 | iter 3050950 step 610190 | loss train: 2.842, val: 2.861 | iter time: 111.03 ms (step) remaining time: 0:01:27
Epoch 34 | iter 3051000 step 610200 | loss train: 2.857, val: 2.861 | iter time: 111.49 ms (step) remaining time: 0:01:22
Epoch 34 | iter 3051050 step 610210 | loss train: 2.895, val: 2.861 | iter time: 111.24 ms (step) remaining time: 0:01:16
Epoch 34 | iter 3051100 step 610220 | loss train: 2.843, val: 2.861 | iter time: 118.35 ms (step) remaining time: 0:01:11
Epoch 34 | iter 3051150 step 610230 | loss train: 2.748, val: 2.861 | iter time: 117.55 ms (step) remaining time: 0:01:05
Epoch 34 | iter 3051200 step 610240 | loss train: 2.818, val: 2.861 | iter time: 120.08 ms (step) remaining time: 0:01:00
Epoch 34 | iter 3051250 step 610250 | loss train: 2.796, val: 2.861 | iter time: 111.08 ms (step) remaining time: 0:00:54
Epoch 34 | iter 3051300 step 610260 | loss train: 2.813, val: 2.861 | iter time: 111.14 ms (step) remaining time: 0:00:49
Epoch 34 | iter 3051350 step 610270 | loss train: 2.823, val: 2.861 | iter time: 118.23 ms (step) remaining time: 0:00:44
Epoch 34 | iter 3051400 step 610280 | loss train: 2.777, val: 2.861 | iter time: 110.98 ms (step) remaining time: 0:00:38
Epoch 34 | iter 3051450 step 610290 | loss train: 2.763, val: 2.861 | iter time: 111.19 ms (step) remaining time: 0:00:33
Epoch 34 | iter 3051500 step 610300 | loss train: 2.743, val: 2.861 | iter time: 111.39 ms (step) remaining time: 0:00:27
Epoch 34 | iter 3051550 step 610310 | loss train: 2.830, val: 2.861 | iter time: 116.14 ms (step) remaining time: 0:00:22
Epoch 34 | iter 3051600 step 610320 | loss train: 2.753, val: 2.861 | iter time: 111.16 ms (step) remaining time: 0:00:17
Epoch 34 | iter 3051650 step 610330 | loss train: 2.819, val: 2.861 | iter time: 111.41 ms (step) remaining time: 0:00:11
Epoch 34 | iter 3051700 step 610340 | loss train: 2.769, val: 2.861 | iter time: 111.13 ms (step) remaining time: 0:00:06
Epoch 34 | iter 3051750 step 610350 | loss train: 2.834, val: 2.861 | iter time: 111.30 ms (step) remaining time: 0:00:00
Saving checkpoint to 'out/pretrain/nano-gpt/final/lit_model.pth'
Training time: 330925.37s
Memory used: 13.27 GB
```

# Training speed

The training speed matches the reported ~4 days in nanoGPT. However, GPU utilization was not maximized and MFU was only 31%. While we could have changed hyperparameters (micro batch size, torch compile settings, etc.) to increase memory pressure and overall utilization, I decided to keep the settings the same for a closer comparison.

- 3.83 days
- ~920k tokens/sec
- ~31% MFU

# Discussion

The final validation loss is ~2.87, which lower than the one reported in [nanoGPT (~3.12)](https://github.com/karpathy/nanoGPT/blob/master/README.md#reproducing-gpt-2). Minor differences in the model architecture as explained above are likely the explanation for this difference. As seen above, the validation loss curve has higher variance than the image shown on the nanoGPT README. This could be explained due to me mistakenly configuring 100 samples for validation, whereas [nanoGPT used 200](https://github.com/karpathy/nanoGPT/blob/9755682b981a45507f6eb9b11eadef8cb83cebd5/config/train_gpt2.py#L21). In addition, nanoGPT did [sampling over the entire validation dataset](https://github.com/karpathy/nanoGPT/blob/9755682b981a45507f6eb9b11eadef8cb83cebd5/train.py#L222) whereas the LitGPT code reuses the same samples for every validation run (no shuffling in data loader). Despite this, the convergence rate matches the original plot well.

Personal note: While this is not a ground breaking result by any means, I hope this eliminates some of your doubts that LitGPT's pretraining script does anything fundamentally wrong. Please use these results as future reference. I also encourage anyone to repeat this experiment on larger models and share your findings!

cc @rasbt @Andrei-Aksionov @lantiga

**Comment by rasbt** [comment]
This is great! Thanks for running this experiment @awaelchli 

We could include this write-up and config file in the examples folder: https://github.com/Lightning-AI/litgpt/tree/main/tutorials/examples (e.g., "gpt2-reproduction" or something like this)

**Comment by rasbt** [comment]
CC @sanyalsunny111

**Comment by pytholic** [comment]
Hi @awaelchli @rasbt . I was wondering if this is the right approach to utilize LitGPT for new custom models/data that do not currently exist in the framework.

I am working with scratch implementation, for now, to do some benchmarking with GPT-2 and [SmolLM](https://huggingface.co/collections/HuggingFaceTB/smollm-6695016cad7167254ce15966) family. However, I would like to switch to LitGPT now and continue it there.

**Comment by lijinhan21** [comment]
hi @awaelchli, thanks for sharing this amazing results! I was trying to reproduce SmolLM2 using litgpt, and found that although training loss and validation loss look reasonable, the benchmark scores are all random guessing. I'm wondering if you have tested your reproduction of gpt2 on any benchmark? Are they reasonable?
